2015/proposals/banaszkiewicz-piotr.md

```diff
@@ -0,0 +1,158 @@
+# Installation-test scripts: aggregation server, script enhancements
+
+## Abstract
+
+While Software Carpentry workshops gather a lot of feedback regarding
+instructors' teaching, not much attention was drawn to students' computers and
+their issues with installation of open source software used during workshops.
+
+This project will focus on bringing a working server for aggregating of
+Software Carpentry installation testing scripts with additional goal of
+providing easier and better experience for students using the script itself.
+
+## Features
+
+The server should be able to aggregate incoming (via REST API) data from
+installation testing script.
+
+The server should allow Software Carpentry admins to view statistics from
+uploaded data.
+
+Installation script should upload diagnostic data consisting of:
+
+* operating system information
+* installed packages and their versions
+* failed packages checks and error messanges
+
+Additionally, the data may be matched against workshop
+[reference package list](https://github.com/wking/swc-setup-installation-test/issues/2).
+
+## Technical Details
+
+The server will be implemented using [Django](https://www.djangoproject.com/)
+and [SQLite](https://docs.djangoproject.com/en/1.7/ref/databases/) database
+for lower footprint and easier maintainance (and because I feel comfortable
+using them).
+
+API, in a RESTful fashion, will be implemented using
+[django-tastypie](http://tastypieapi.org/).
+
+Rough database structure:
+
+* system information table
+    * operating system family (Windows / MacOSX / Linux / Other)
+    * OS version (8.1 / 10.9 / Ubuntu / ?)
+    * Hardware architecture (x86_64 / x86 / Other)
+* package checks table
+    * package name
+    * requested package version
+    * found package version
+    * failed check?
+    * fail reason
+
+Every entry in both tables will be additionally assigned a
+[universally unique identifier](http://en.wikipedia.org/wiki/Universally_unique_identifier).
+Every single UUID will correspond to one diagnostic data upload.
```

Comment:
On Thu, Mar 19, 2015 at 11:03:38AM -0700, Piotr Banaszkiewicz wrote:
> 1) One run of installation testing script will produce 1 entry to
> the system information table and possibly many entries to the
> package checks table. Entries in both tables won't share the same ID
> (ie. primary key) because of a) two tables engaged, and b) auto
> increment.

That's what foreign keys are for [1,2], and if linking from the
package-check table to the system-info table is all you need, using
auto-incremented integer IDs will be easier and faster than using
UUIDs.

[1]: https://www.sqlite.org/foreignkeys.html
[2]: http://www.postgresql.org/docs/9.4/static/tutorial-fk.html